enzyme/Enzyme/EnzymeLogic.cpp

```diff
@@ -3390,6 +3389,349 @@ Function *EnzymeLogic::CreatePrimalAndGradient(
+      dretAlloca);
+
+  for (BasicBlock &oBB : *gutils->oldFunc) {
+    // Don't create derivatives for code that results in termination
+    if (guaranteedUnreachable.find(&oBB) != guaranteedUnreachable.end()) {
+      auto newBB = cast<BasicBlock>(gutils->getNewFromOriginal(&oBB));
+      std::vector<BasicBlock *> toRemove;
+      if (auto II = dyn_cast<InvokeInst>(oBB.getTerminator())) {
+        toRemove.push_back(
+            cast<BasicBlock>(gutils->getNewFromOriginal(II->getNormalDest())));
+      } else {
+        for (auto next : successors(&oBB)) {
+          auto sucBB = cast<BasicBlock>(gutils->getNewFromOriginal(next));
+          toRemove.push_back(sucBB);
+        }
+      }
+
+      for (auto sucBB : toRemove) {
+        sucBB->removePredecessor(newBB);
+      }
+
+      std::vector<Instruction *> toerase;
+      for (auto &I : oBB) {
+        toerase.push_back(&I);
+      }
+      for (auto I : toerase) {
+        maker.eraseIfUnused(*I, /*erase*/ true, /*check*/ true);
+      }
+      if (newBB->getTerminator())
+        newBB->getTerminator()->eraseFromParent();
+      IRBuilder<> builder(newBB);
+      builder.CreateUnreachable();
+      continue;
+    }
+
+    auto term = oBB.getTerminator();
+    assert(term);
+    if (!isa<ReturnInst>(term) && !isa<BranchInst>(term) &&
+        !isa<SwitchInst>(term)) {
+      llvm::errs() << *oBB.getParent() << "\n";
+      llvm::errs() << "unknown terminator instance " << *term << "\n";
+      assert(0 && "unknown terminator inst");
+    }
+
+    auto first = oBB.begin();
+    auto last = oBB.empty() ? oBB.end() : std::prev(oBB.end());
+    for (auto it = first; it != last; ++it) {
+      maker.visit(&*it);
+    }
+
+    createTerminator(gutils, constant_args, &oBB, retAlloca, dretAlloca,
+                     retType);
+  }
+
+  gutils->eraseFictiousPHIs();
+
+  BasicBlock *entry = &gutils->newFunc->getEntryBlock();
+
+  auto Arch =
+      llvm::Triple(gutils->newFunc->getParent()->getTargetTriple()).getArch();
+
+  while (gutils->inversionAllocs->size() > 0) {
+    Instruction *inst = &gutils->inversionAllocs->back();
+    if (isa<AllocaInst>(inst))
+      inst->moveBefore(&gutils->newFunc->getEntryBlock().front());
+    else
+      inst->moveBefore(entry->getFirstNonPHIOrDbgOrLifetime());
+  }
+
+  (IRBuilder<>(gutils->inversionAllocs)).CreateUnreachable();
+  DeleteDeadBlock(gutils->inversionAllocs);
+  for (auto BBs : gutils->reverseBlocks) {
+    if (pred_begin(BBs.second.front()) == pred_end(BBs.second.front())) {
+      (IRBuilder<>(BBs.second.front())).CreateUnreachable();
+      DeleteDeadBlock(BBs.second.front());
+    }
+  }
+
+  for (Argument &Arg : gutils->newFunc->args()) {
+    if (Arg.hasAttribute(Attribute::Returned))
+      Arg.removeAttr(Attribute::Returned);
+    if (Arg.hasAttribute(Attribute::StructRet))
+      Arg.removeAttr(Attribute::StructRet);
+  }
+  if (gutils->newFunc->hasFnAttribute(Attribute::OptimizeNone))
```

Comment:
While we're cleaning up, can we move the function attribute code into a common function used by all methods.

Reply:
The function attribute code for CreateAugmentedPrimal is not quite the same, so ill leave it out.

Reply:
How precisely do they differ?